doc/man1/linxcfg.1
.TH LINXCFG 1 "2010-01-20" 1.5 "LINX"
.SH NAME
linxcfg \- a configuration utility for creating LINX links to other nodes
.SH SYNOPSIS
.B linxcfg [-t <cm>] [-v] <command> [COMMAND PARAMETERS and OPTIONS]

.B linxcfg [-t <cm>] [-v] destroy <linkname> [<link 2>...]

.B linxcfg -h

.B linxcfg help <cm>

.B linxcfg -t <cm> [help]

.SH DESCRIPTION
.B linxcfg
is the LINX configuration utility. Depending on the
.I command
given,
.B linxcfg
can destroy or create links using the specified Connection Manager (CM).

.SH OPTIONS

.B linxcfg
has common options to all commands which should be placed
.B before
the actual
.I command

.IP "-h, --help"
Display help and exit.

.IP "-t, --type <cm>"
Specifies
.I cm
as the Connection Manager to use. If this option is not provided,
.I "eth"
will be used as the default CM.

You can use
.B -h
to list all available CMs.

.IP "-v, --verbose"
Verbose mode.

.IP "-c, --cf, --conn-feature <feature[=value]>"
Adds a connection feature. This option can be specified several times.

.IP "-l, --lf, --link-feature <feature[=value]>"
Adds a link feature. This option can be specified several times.

.SH COMMANDS

.TP
create PARAMETERS
Creates a connection link to another node. The
.I PARAMETERS
depend on the CM specified, see below for more details.

A remote link cannot be used until also the remote node creates
the other end of the link (a link back to the originating node).
The state is "connecting" until both sides are ready and
communication via the LINX protocol can begin.


.TP
destroy <linkname> [<link2>...]
Destroy the connection link with the name
.I linkname.

.TP
help [<cm>]
Provides help about the given
.I cm
, either specified in
.B -t
or after the
.B help
command.

If no
.I cm
was specified, it will show a list of available CMs.

.SH OPTIONS TO DIFFERENT CONNECTION MANAGERS

.SS Ethernet CM

The
.B Ethernet CM
is specified with the
.I "eth"
keyword, which is actually the default if no other CM was specified.

The
.B create
command takes the following options:

.TP
create <macaddr> <interface> <linkname> [OPTIONAL PARAMETERS]

.TP
macaddr
Specify the remote mac address of the node to connect to, e.g.
0a:1b:2c:3d:4d:5e.

.TP
interface
Specify the Ethernet interface to use. Ex.: eth0.

.TP
linkname
Name of the connection link. This is the link part of the path in a remote
hunt name to LINX endpoints on the other node. It will also be the
name seen in
.I /proc/net/linx/cm/eth/<linkname>

.IP "Optional parameters"

.TP
--window_size=nnn
This is the send/receive window_size, in number of packets on the link,
and may need to be modified to adapt LINX to really slow or really fast
Ethernet performance.
.br
Both the local and remote side negotiate the window size, being the
smallest of the two sizes the selected one.
The default 128 messages should be sufficient in most configurations.
The window size shall always be of a power of 2.
Size 0 means to use the default window size.

.TP
--defer_queue_size=nnn
The defer queue size is in packages, with the size depending on the LINX link.
The defer queue is used on the sender side when the send queue is full.
Every message sent, when the send queue is full, is stored in the defer queue,
until the send queue has room for more messages.
The defer queue size is the maximum number of packages, with messages,
that can be stored in the defer queue before the link is disconnected,
due to lack or resources.
The default value 2048 packages should be sufficient in most systems.
Size 0 means to use the default defer queue size.

.TP
--conn_tmo=nnn
The connect timeout specifies the time, in milliseconds, to wait until a
connection attempt fails and a new attempt is made. Once connected, it is used
as a supervision time-out, i.e. every conn_tmo/3 milliseconds, LINX checks if
any packets from the peer has been received. LINX disconnects the connection
after four consequtive conn_tmo/3 timeout periods without any packets from the
peer. Tmo 0 means to use the default timeout (1000ms).

.TP
--mtu=nnn
The MTU (Maximum Transmission Unit) specifies the size in  bytes of the
largets packet that the Ethernet protocol can pass onwards (this excludes the
size of the Ethernet header). If not set the MTU is fetched from the interface.
Typcially a MTU of 1500 bytes is used for Ethernet.

.TP
--attributes=s
The attribute option gives the possibility to assign an arbitrary string to
the link when it is created. This string is included in the new_link signal,
which is sent to all link supervisors. More information on link
supervisors can be found in
.TP
--coreid=nn
The coreid is used in a multicore environment to configure the 
destination coreid.

.BR linx "(7) "
man page.

.SS TCP CM

The TCP CM is specified using the
.I tcp
keyword.

The
.B create
command takes the following parameters:

create <ip> <linkname> [OPTIONAL PARAMETERS]

.TP
ip
The IP address to connect to.

.TP
linkname
Name of the connection link. This is the LINX link part of the path in
a remote hunt name to LINX endpoints on the other node. It will also be the
name seen in
.I /proc/net/linx/cm/tcp/<linkname>

.IP "Optional parameters"

.TP
--live_tmo=<size>
The live_tmo parameter is the time in milliseconds between every heartbeat
that is used to detect if the connection has gone down. The default value is
1000 ms.
.TP
--use_nagle=<bool>
Set to 1 if nagle algorithm shall be used on the socket for the connection.
Default is off.
.TP
--attributes=<s>
The attribute option gives the possibility to assign an arbitrary string to
the link when it is created. This string is included in the new_link signal,
which is sent to all link supervisors.

.SS RIO CM

The RIO CM is specified using the
.I rio
keyword.

The
.B create
command takes the following parameters:

create <local_port> <port> <dev_id> <mbox> <if> <link> [OPTIONAL PARAMETERS]

.TP
local_port
The local RapidIO port to connect to.

.TP
port
The remote RapidIO port to connect to.

.TP
dev_id
The RapidIO device to connect to.

.TP
mbox
The RapidIO mailbox id to use.

.TP
if
The RapidIO interface to use.

.TP
link
Name of the connection link. This is the LINX link part of the path in
a remote hunt name to LINX endpoints on the other node. It will also be the
name seen in
.I /proc/net/linx/rlnh

.IP "Optional parameters"

.TP
--tmo=<tmo>
The tmo parameter is the time in hundreds of milliseconds between
every heartbeat that is used to detect if the connection has gone
down. The default value is 500 ms.
.TP
--mtu=<mtu>
Specifies the MTU (Maximum Transmission Unit) in bytes of the largest
that the RIO CM can pass onwards, including the size of the RapidIO
header. If not explicitly given, the MTU is fetched from the RapidIO device.

.SH FILES
None.
.SH DIAGNOSTICS
.I linxcfg
will display more verbose information to standard out, if the -v option is
specified.
.SH KNOWN BUGS
None.
.SH EXAMPLES
.nf
linxcfg create 01:23:a4:4f:b3:ac eth0 link_A

linxcfg destroy link_A

linxcfg -t tcp create 192.168.1.1 link_A

linxcfg -t tcp destroy link_A

linxcfg -t rio create 0 0 0 0 rio0 riolink0 --mtu=128 --tmo=10

linxcfg -t rio destroy riolink0

.fi
.SH "SEE ALSO"
.BR linx "(7), "
.BR linxstat "(1), "
.BR linxdisc "(8)"

.SH AUTHOR
Enea LINX team

.SH COPYRIGHT

Copyright (c) 2006-2010, Enea Software AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
Neither the name of Enea Software AB nor the names of its
contributors may be used to endorse or promote products derived from this
software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


ChangeLog
Version 2.5.1 (2011-07-08)
==========
* The limit of number of sockets per link allowed lowered from 1024 to two.
* Removed a check in LINX kernel module for max signal size.
* Fixes for the multicore header in the ECM and Wireshark patch.
		
Version 2.5.0 (2011-04-26)
==========
* Fixed a bug in linxstat.
* Updated manpage for mkethcon and help info for mkethcon and linxcfg to reflect
  reality.
* libcfg was updated to support cmcl connections.
* Fixed minor bugs in rm*con binaries.
* Fixed build errors for various kernel versions. ipv6 for <2.6.22 and db.c for
  newer ones >2.6.36.
* Fixed some issues in the linx Gateway related to LINXCR-1397
* Added support for ipv6 in the tcp connection manager.
* Improved the performance of intra-process communication with pointer passing.
* Fixed some issues found by coverity.
* Removed some warnings when building linxstat.
* Remove all BUG BUG_ON WARN_ON from linx modules to avoid using __FILE__
  define in the errors. This define causes the linx module to be different
  if compiled in different folder.
* Added fixes in the ECM were a malformed received packet could lead to
  kernel panic.
* Fixed issue with multi-access to LINX sockets from user-space
  applications could lead to kernel panic.
* Added new features to linxstat: print info only for a given huntname
  and display only a specified number of characters from huntname
* Added coreid header
* Added support for 2.5.35 kernels.
* Fixed bug with negative signal filter.
* Fixed bug with missing internal signals in the message trace.
* Added patch file for Ethernet CM to compensate for faulty Ethernet
  drivers that don't allocate enough headroom in skb when using VLAN.
* Added support for 2.6.33 and 2.6.34 kernels.
* Fixed vlan bug reproducing on certain machines
* Fixed bug with receive_w_tmo where tmo=0 resulted in no signal above
  4095 could be received.
* Fixed string handling bug in linxgws.
* Fixed memory allocation issue in linxgws.
* Fixed a bug in linxdisc regarding several interfaces.
* Added ip config in linxnet conf file
* Linxstat now uses a function interface to retrieve info instead of
ioctl.
	
Version 2.4.0 (2010-02-22)
==========	
* Added missing riocm.txt to MANIFEST
* Added RIOCM documentation. net/linx/riocm/riocm.txt
* Updated RELEASE_NOTES for 2.4.0 release.
* Added support for creating links using RapidIO CM connections in linxcfg.

Version 2.4.0-beta (2010-01-15)
==========
* Added a RapidIO connection manager.
* Remove debug printk in device notification handler.

Version 2.3.0 (2009-12-9)
==========
* Fixed build errors for vanilla kernel 2.6.32
* Example conf file for Linxdisc now emphasizes that VLAN interfaces are supported.
* RLNH control data is now always sent from kernel context.
* Altered the keepalive mechanism for the Ethernet CM. The old implementation
  was sensitive to packet loss. Contributed by paull7 @ sourceforge. Note: The
  new implementation is backwards compatible.
* Added support for 2.6.31 kernels.
* Fix LINX_IOCTL_INFO ioctl request for gathering process' receive filters.
* Document LINX_INFO_RECV_QUEUE_2.
* Drop tcp connect packets that doesn't match the current protocol
version.
* Add hardware address size in ethernet connection header.
* Man page for linx_request_new_link and linx_cancel_new_link was missing
* Fixed bug where the request ack bit was not set when the sliding window queue
  is becoming full.
* Fixed memory leaks in the code for create and destroy connections and links.	
* New shared memory connection manager.
* Wireshark plugin now supports LINX TCP Connection manager.
* Wrong RLNH protocol version was listed in documentation.	
	
Version 2.2.0 (2009-03-31)
==========
* New Linxdisc and Linx internal messages patches for Wireshark 1.0.6
* New LINX patches for tcpdump 4.0.0 and libpcap 1.0.0
* Trace of internal LINX messages now use official LINX ethernet type, 0x8911.
* Added new function linx_get_owner(), return the pid of the owner of a LINX endpoint.	
* Added missing files to MANIFEST.
* Example files now have the correct copyright header.
* Fixed oob-issue where the a link in some cases wasn't created.
* Removed a state in the tcp cm state machine.
* Did some major cleanup in tcp cm.
* Removed a kernel function from af_linx.c
* Fixed a bug in tcp cm regarding halt of workqueue when connecting to dead peers
* Linux Gateway Server and Client added to distribution, also gwcmd tool to test Linx Gateway Servers. 
* Removed bug_on's and fixed better error reporting back to user space on failed send.
* Fix bug in linx_info_filter_32
* Fixed alignment issue in linxstat.
* Removed linx_create/destroy_eth/tcp_link functions.
* Avoid calling kmalloc(0) from Linx IPC-layer.

Version 2.1.0 (2008-11-10)
==========	
* LINX now runs on x86 2.6.27 Linux kernel.
* Add support for mips(64).
* Add patches to libpcap and tcpdump configure scripts for Octeon.
* The TCP CM now support OOB messaging.
* Fixes for running w errorchecks enabled.
* Improved garbage collection of sockets in the tcp_cm.
* Minor changes of linxstat output.
* The linxstat program can now tell if a signal in the receive queue is OOB.
* Support for allocating fragmentated skb's is now a LINX kernel module param and no longer
  a comple-time option.
* Fixed bug when building the LINX kernel module with memory fragmentation support on and
  sending large unpaged signals.
* Fixed bug in LINX message trace. If a fragmented message was received, only the first
  fragment was traced.
* Added optimized linx_do_sendmsg call for attach/hunt/tmo/newlink signals.
* Pulled back warning level a bit to avoid complaints from kernel headers.
* Fixed bug that lead to the LINX kernel module couldn't be unloaded.
* linx_skb_copy_bits removed
* linx_ioctl_info_*() functions moved to a seperate file.
* Internal handling of attach/tmo/newlink signals is simplified.
* Updated liblinx to user LINX_IOCTL_SEND and LINX_IOCTL_RECEIVE instead of sendto/recvfrom/sendmsg

linxcfg/linxcfg.c
/*
 * Copyright (c) 2007, Enea Software AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of Enea Software AB nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdarg.h>
#define _GNU_SOURCE
#include <getopt.h>

#include "config.h"
#include "linx_cfg.h"
#include "linx_cfg_cm.h"

#define _usage   linxcfg_usage
#define _verbose linxcfg_verbose

static unsigned int VERBOSE = 0;
static unsigned int DO_HELP = 0;

#define TOTAL_MAIN_OPTIONS 7
static struct option main_options[] = {
	{"link-feature", 1, 0, 'l'},
	{"lf", 1, 0, 'l'},
	{"conn-feature", 1, 0, 'c'},
	{"cf", 1, 0, 'c'},
	{"help", 0, 0, 'h'},
	{"type", 1, 0, 't'},
	{"verbose", 0, 0, 'v'}
};
static char *main_short_options = "c:ht:v";

char *conn_fstr = NULL;
char *link_fstr = NULL;

/* this returns */
unsigned int linxcfg_verbose(const char *fmt, ...)
{
	va_list ap;

	/* Do not proceed if not in verbose mode */
	if (!VERBOSE) {
		return 0;
	}

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);

	return 1;
}

void linxcfg_usage(int indent, const char *fmt, ...)
{
	int i;
	va_list ap;

	/* lame indent */
	for (i = 0; i < indent; ++i) {
		printf(" ");
	}

	va_start(ap, fmt);
	vprintf(fmt, ap);
	va_end(ap);
}

void *Malloc(size_t s)
{
	void *ret;

	if ((ret = malloc(s)) == NULL) {
		perror("malloc");
		exit(1);
	}

	return ret;
}

void show_help(const char *cm_name)
{
	int i = 0;

	_usage(0,
	       "Usage: linxcfg [OPTIONS] <command> [COMMAND PARAMETERS]\n\n");
	_usage(4, "OPTIONS:\n");
	_usage(8, "-h      Shows this help message and exits.\n");
	_usage(8,
	       "-t <cm> Use the given <cm> for <command> "
	       "(default is 'eth')\n");
	_usage(8, "-v      Sets verbose mode on.\n\n");
	_usage(4, "FEATURES:\n");
	_usage(8,
	       "--conn-feature, --cf, -c <f[=v]>  "
	       "Adds a connection feature.\n");
	_usage(8,
	       "--link-feature, --lf, -l <f[=v]>  " "Adds a link feature.\n");
	_usage(8, "Where f is a feature and v its value (optional),\n");
	_usage(8, "depending on the CM and connection layer versions.\n\n");
	_usage(4, "COMMANDS:\n");
	_usage(8,
	       "create [CM PARAMETERS]  Creates a link using the "
	       "desired CM.\n");
	_usage(8, "destroy <link>          Destroys the given link.\n");
	_usage(8, "help [<cm>]             Provides help about cm.\n\n");

	if (cm_name) {
		/* find out the position of the desired cm */
		while (cm_handler_list[i] != NULL) {
			if (!strncmp(cm_handler_list[i]->name,
				     cm_name, strlen(cm_name))) {
				break;
			}
			i++;
		}

		if (cm_handler_list[i] != NULL) {
			_usage(4, "CM PARAMETERS and OPTIONS for '%s':\n",
			       cm_name);
			_usage(8, "%s (%s)\n",
			       cm_handler_list[i]->description,
			       cm_handler_list[i]->name);
			if (cm_handler_list[i]->help != NULL) {
				cm_handler_list[i]->help(12);
			}
			return;
		}
	}

	_usage(4, "AVAILABLE CM TYPES:\n");
	for (i = 0; cm_handler_list[i] != NULL; i++) {
		_usage(8,
		       "%s\t%s\n",
		       cm_handler_list[i]->name,
		       cm_handler_list[i]->description);
	}
}

int get_command(const char *command)
{
	if (command == NULL)
		return CMD_UNKNOWN;

	if (strncmp(command, "create", 6) == 0 && command[6] == '\0') {
		return CMD_CREATE;
	} else if (strncmp(command, "destroy", 7) == 0 && command[7] == '\0') {
		return CMD_DESTROY;
	} else if (strncmp(command, "help", 4) == 0 && command[4] == '\0') {
		return CMD_HELP;
	}

	return CMD_UNKNOWN;
}

int get_all_longoptions(struct option **ops)
{
	char **list = NULL;
	int x;
	int total = TOTAL_MAIN_OPTIONS;
	struct option *new;

	*ops = Malloc(sizeof(struct option) * TOTAL_MAIN_OPTIONS);
	memcpy(*ops, main_options, sizeof(struct option) * TOTAL_MAIN_OPTIONS);

	for (x = 0; cm_handler_list[x] != NULL; x++) {
		int i;
		int count = config_get_option_list(cm_handler_list[x]->t,
						   &list,
						   0);

		if (count <= 0) {
			fprintf(stderr, "%s didn't return a list\n",
				cm_handler_list[x]->name);
			exit(1);
		}
		new = realloc(*ops, sizeof(struct option) * (count + total));
		if (new == NULL) {
			perror("realloc");
			exit(1);
		}
		for (i = 0; i < count; i++) {
			new[total + i].name = list[i];
			new[total + i].has_arg = 1;
			new[total + i].flag = NULL;
			new[total + i].val = 0;
		}
		free(list);	/* notice we don't free the strings,
				 * they are still referenced in ops
				 */
		*ops = new;
		total += count;
	}
	new = realloc(*ops, sizeof(struct option) * (total + 1));
	if (new == NULL) {
		perror("realloc");
		exit(1);
	}
	new[total].name = NULL;
	new[total].has_arg = 0;
	new[total].flag = NULL;
	new[total].val = 0;

	return total;
}

void add_feature(char **str, char *arg)
{
	char *tmp;

	if (arg == NULL || strlen(arg) == 0) {
		fprintf(stderr, "Error: No feature given.\n");
		return;
	}

	/* allow to have --link-feature blah=bleg or --lf blah:bleh */
	if ((tmp = strchr(arg, '=')) != NULL) {
		*tmp = ':';
	}

	if (*str != NULL) {
		tmp = Malloc(strlen(*str) + strlen(arg) + 2);
		sprintf(tmp, "%s,%s", *str, arg);
		free(*str);
	} else {
		tmp = Malloc(strlen(arg) + 1);
		strcpy(tmp, arg);
	}
	*str = tmp;
}

char *get_link_feature_string(void)
{
	return link_fstr;
}

char *get_connection_feature_string(void)
{
	return conn_fstr;
}

int main(int argc, char **argv)
{
	int x = 0, i, opt, ind;
	struct option_value *opval;
	linxcfg_cm_handler *cmh;
	char *cm_name = NULL;

	struct option *lops;

	x = get_all_longoptions(&lops);
	if (x <= 0) {
		fprintf(stderr, "Could not get options (%d)\n", x);
		exit(1);
	}

	opval = Malloc(sizeof(struct option_value) * (x + 1));
	i = 0;			/* keeps track on the number of options set */
	while (1) {
		opt = getopt_long(argc, argv, main_short_options, lops, &ind);
		if (opt == -1)
			break;
		switch (opt) {
		case 0:
			/* --option=value kind of option */
			opval[i].name = Malloc(strlen(lops[ind].name) + 1);
			strcpy(opval[i].name, lops[ind].name);
			if (optarg) {
				opval[i].val = Malloc(sizeof(optarg + 1));
				strcpy(opval[i].val, optarg);
			} else {
				opval[i].val = NULL;
			}
			i++;
			break;
		case 'c':
			/* connection feature */
			add_feature(&conn_fstr, optarg);
			break;
		case 'l':
			/* link feature */
			add_feature(&link_fstr, optarg);
			break;
		case 't':
			/* specifies the type of CM */
			cm_name = Malloc(strlen(optarg) + 1);
			strcpy(cm_name, optarg);
			break;
		case 'h':
			/* help */
			DO_HELP = 1;
			break;
		case 'v':
			/* verbose */
			VERBOSE = 1;
			break;
		case '?':
			printf("unknown option code: %d\n", opt);
			show_help(NULL);
			exit(1);
			break;
		default:
			printf("?? opt code 0%o ??\n", opt);
			show_help(NULL);
			exit(1);
		}
	}
	opval[i].name = NULL;

	if (DO_HELP) {
		if (cm_name == NULL && optind < argc)
			cm_name = argv[optind];
		show_help(cm_name);
		exit(0);
	}

	if (optind >= argc) {
		fprintf(stderr, "Too few arguments\n");
		show_help(cm_name);
		exit(1);
	}

	/* if none given, set default <cm> */
	if (cm_name == NULL) {
		/* it could be the 'help' command */
		if (get_command(argv[optind]) == CMD_HELP) {
			optind++;
			if (optind < argc)
				cm_name = argv[optind];
			show_help(cm_name);
			exit(0);
		}
		_verbose("Using " DEFAULT_CM " as default CM.\n");
		cm_name = (char *)Malloc(strlen(DEFAULT_CM) + 1);
		strcpy(cm_name, DEFAULT_CM);
	}

	if (cm_name == NULL) {
		fprintf(stderr, "No default CM found!\n");
		show_help(NULL);
		exit(1);
	}

	x = 0;
	for (x = 0; cm_handler_list[x] != NULL; x++) {
		cmh = cm_handler_list[x];
		if (cmh->name == NULL)
			continue;
		if (cmh->name != NULL &&
		    cm_name != NULL &&
		    strncmp(cm_name, cmh->name, strlen(cmh->name)) == 0 &&
		    cm_name[strlen(cmh->name)] == '\0') {
			_verbose("--- Selected CM is %s\n", cmh->name, opval);
			/* set values */
			for (i = 0; opval[i].name != NULL; i++) {
				if (config_set_value(cmh->t,
						     opval[i].name,
						     opval[i].val)) {
					fprintf(stderr,
						"Wrong value (%s) for %s\n",
						opval[i].val, opval[i].name);
					exit(1);
				}
			}
			i = optind ? optind : 1;
			config_parse_noops(cmh->t, argc - i, &argv[i]);
			return (cmh->init(argc - i, &argv[i]));
		}
	}

	fprintf(stderr, "No such CM: %s\n", cm_name);
	show_help(NULL);

	return 0;
}
